# Austin, Travis County Zillow Data Analysis

Austin, TX is a bustling city with a very attractive economy. Home to a large state institution, the University of Texas, and massive events every year such as Austin City Limits (ACL) and the Formula 1 U.S.A. Grand Prix, Austin is a very attractive city for potential real estate investors, for both appreciation and passive income. Using data obtained from Zillow, we are analyzing to see if numpy can be used to accurately model and so we can extract some predictive capability from said data. This can ultimately help real estate investors better predict the price of potential homes based on the zip code in which they are looking. This analysis will either confirm or deny the ability to model the data and proceed to use it in a predictive manner. 

## Data Import, Processing and Visualization

In [57]:
# Import statements

import pandas as pd
from pathlib import Path
import hvplot.pandas
import numpy as np
import math

In [58]:
# Import Zillow dataset into a pandas dataframe and display said data

zillow_df = pd.read_csv(Path('Resources/zillow_home_values.csv'),
                        infer_datetime_format=True)

display(zillow_df.head())

RegionID  SizeRank  RegionName RegionType StateName State      City  \
0     61639         0       10025        Zip        NY    NY  New York   
1     84654         1       60657        Zip        IL    IL   Chicago   
2     61637         2       10023        Zip        NY    NY  New York   
3     91982         3       77494        Zip        TX    TX      Katy   
4     84616         4       60614        Zip        IL    IL   Chicago   

                              Metro       CountyName  01/31/2000  ...  \
0       New York-Newark-Jersey City  New York County    324383.0  ...   
1          Chicago-Naperville-Elgin      Cook County    313500.0  ...   
2       New York-Newark-Jersey City  New York County    506994.0  ...   
3  Houston-The Woodlands-Sugar Land    Harris County    224938.0  ...   
4          Chicago-Naperville-Elgin      Cook County    409547.0  ...   

   09/30/2021  10/31/2021  11/30/2021  12/31/2021  01/31/2022  02/28/2022  \
0   1132221.0   1136610.0   1135991.0   1136561.0   1142452.0   1154203.0   
1    524133.0    526722.0    529521.0    531520.0    532909.0    533123.0   
2   1498926.0   1507590.0   1511855.0   1512170.0   1522793.0   1532550.0   
3    421500.0    429474.0    436571.0    442317.0    448423.0    457973.0   
4    668926.0    672856.0    677523.0    680459.0    682380.0    682141.0   

   03/31/2022  04/30/2022  05/31/2022  06/30/2022  
0   1172911.0   1189572.0   1203675.0   1209322.0  
1    535804.0    537898.0    542946.0    546766.0  
2   1549563.0   1562744.0   1581340.0   1590483.0  
3    469673.0    483460.0    494924.0    502464.0  
4    685249.0    687700.0    693521.0    698444.0  

[5 rows x 279 columns]

In [59]:
# Splice for only Austin data

zillow_austin = zillow_df.loc[zillow_df['Metro'] == 'Austin-Round Rock']
zillow_austin_df = zillow_austin.loc[zillow_austin['City'] == 'Austin']
display(zillow_austin_df)

RegionID  SizeRank  RegionName RegionType StateName State    City  \
70        92617        71       78704        Zip        TX    TX  Austin   
92        92654        93       78745        Zip        TX    TX  Austin   
443       92657       446       78748        Zip        TX    TX  Austin   
450       92667       453       78758        Zip        TX    TX  Austin   
485       92651       488       78741        Zip        TX    TX  Austin   
520       92662       523       78753        Zip        TX    TX  Austin   
594       92668       597       78759        Zip        TX    TX  Austin   
1604      92653      1617       78744        Zip        TX    TX  Austin   
2060      92634      2075       78723        Zip        TX    TX  Austin   
2157      92658      2172       78749        Zip        TX    TX  Austin   
2404      92640      2421       78729        Zip        TX    TX  Austin   
2639      92628      2661       78717        Zip        TX    TX  Austin   
2780      92638      2801       78727        Zip        TX    TX  Austin   
2983      92642      3007       78731        Zip        TX    TX  Austin   
3378      92618      3411       78705        Zip        TX    TX  Austin   
3508      92639      3542       78728        Zip        TX    TX  Austin   
3700      92655      3735       78746        Zip        TX    TX  Austin   
3772      92616      3809       78703        Zip        TX    TX  Austin   
3957      92659      3993       78750        Zip        TX    TX  Austin   
4132      92615      4172       78702        Zip        TX    TX  Austin   
4218      92666      4260       78757        Zip        TX    TX  Austin   
5118      92646      5168       78735        Zip        TX    TX  Austin   
5288      92663      5338       78754        Zip        TX    TX  Austin   
5397      92660      5450       78751        Zip        TX    TX  Austin   
5638      92661      5693       78752        Zip        TX    TX  Austin   
6437      92635      6499       78724        Zip        TX    TX  Austin   
6792      92656      6859       78747        Zip        TX    TX  Austin   
6809      92555      6876       78617        Zip        TX    TX  Austin   
7048      92648      7119       78737        Zip        TX    TX  Austin   
7279      92650      7354       78739        Zip        TX    TX  Austin   
7323      92614      7401       78701        Zip        TX    TX  Austin   
7402      92637      7479       78726        Zip        TX    TX  Austin   
7445      92643      7526       78732        Zip        TX    TX  Austin   
9247      92665      9358       78756        Zip        TX    TX  Austin   
9515      92632      9629       78721        Zip        TX    TX  Austin   
9728      92641      9846       78730        Zip        TX    TX  Austin   
9902      92647     10021       78736        Zip        TX    TX  Austin   
10839     92636     10974       78725        Zip        TX    TX  Austin   
10942     92644     11082       78733        Zip        TX    TX  Austin   
10987     92633     11128       78722        Zip        TX    TX  Austin   
12423     92585     12602       78652        Zip        TX    TX  Austin   
21376     92630     22577       78719        Zip        TX    TX  Austin   
23803     92652     25635       78742        Zip        TX    TX  Austin   

                   Metro     CountyName  01/31/2000  ...  09/30/2021  \
70     Austin-Round Rock  Travis County    258755.0  ...    950810.0   
92     Austin-Round Rock  Travis County    158425.0  ...    541850.0   
443    Austin-Round Rock  Travis County    168963.0  ...    498905.0   
450    Austin-Round Rock  Travis County    128102.0  ...    497849.0   
485    Austin-Round Rock  Travis County    147654.0  ...    456450.0   
520    Austin-Round Rock  Travis County    129273.0  ...    438699.0   
594    Austin-Round Rock  Travis County    253680.0  ...    763853.0   
1604   Austin-Round Rock  Travis County    120228.0  ...    401862.0   
2060   Austin-

In [60]:
# Create a dataframe with only the dates as the rows and zip codes as the columns with the prices as the primary data

dates_df = zillow_austin_df.drop(columns=['RegionID', 'RegionName','RegionType','State', 'City', 'Metro', 'CountyName', 'SizeRank','StateName']).T
zipcode_df = zillow_austin_df['RegionName']

dates_df.set_index(dates_df.columns[0])
dates_df.columns = zipcode_df

display(dates_df)

RegionName      78704     78745     78748     78758     78741     78753  \
01/31/2000   258755.0  158425.0  168963.0  128102.0  147654.0  129273.0   
02/29/2000   263523.0  158246.0  169557.0  128616.0  149019.0  129699.0   
03/31/2000   268348.0  159585.0  170277.0  129124.0  149457.0  129880.0   
04/30/2000   272660.0  160477.0  171529.0  130212.0  149806.0  130571.0   
05/31/2000   276739.0  160869.0  172626.0  130949.0  149329.0  131124.0   
...               ...       ...       ...       ...       ...       ...   
02/28/2022  1053418.0  589353.0  541130.0  537845.0  496936.0  474259.0   
03/31/2022  1072684.0  602249.0  553220.0  551156.0  508005.0  485392.0   
04/30/2022  1087691.0  612028.0  561694.0  561618.0  515889.0  494108.0   
05/31/2022  1096965.0  618316.0  568101.0  567226.0  519962.0  500128.0   
06/30/2022  1095829.0  613862.0  563042.0  560180.0  514218.0  495392.0   

RegionName     78759     78744     78723     78749  ...     78756     78721  \
01/31/2000  253680.0  120228.0  156793.0  207483.0  ...  296389.0  138992.0   
02/29/2000  255556.0  120856.0  159300.0  206991.0  ...  304238.0  140029.0   
03/31/2000  257186.0  121483.0  164323.0  206529.0  ...  305509.0  142717.0   
04/30/2000  259497.0  122325.0  167621.0  205803.0  ...  310970.0  147868.0   
05/31/2000  260554.0  123007.0  170671.0  205134.0  ...  311430.0  154022.0   
...              ...       ...       ...       ...  ...       ...       ...   
02/28/2022  844218.0  440666.0  652733.0  686617.0  ...  931114.0  559595.0   
03/31/2022  862457.0  451566.0  663550.0  701109.0  ...  944235.0  571327.0   
04/30/2022  876764.0  460693.0  670097.0  713245.0  ...  952069.0  576838.0   
05/31/2022  884352.0  467861.0  672464.0  721249.0  ...  957718.0  579854.0   
06/30/2022  874722.0  464934.0  661809.0  713630.0  ...  950845.0  570825.0   

RegionName      78730     78736     78725      78733     78722     78652  \
01/31/2000   488050.0  223100.0  120213.0   438890.0       NaN  192614.0   
02/29/2000   487684.0  222703.0  120984.0   438804.0       NaN  193082.0   
03/31/2000   489031.0  223886.0  121723.0   439757.0       NaN  194118.0   
04/30/2000   491682.0  224554.0  123021.0   441868.0       NaN  195968.0   
05/31/2000   494572.0  224520.0  124149.0   441903.0       NaN  197784.0   
...               ...       ...       ...        ...       ...       ...   
02/28/2022  1346091.0  721532.0  388065.0  1397156.0  741936.0  595321.0   
03/31/2022  1385862.0  736466.0  398068.0  1431161.0  760522.0  610890.0   
04/30/2022  1421492.0  748991.0  407091.0  1465264.0  773833.0  624577.0   
05/31/2022  1447688.0  757909.0  415376.0  1487102.0  787315.0  632163.0   
06/30/2022  1443527.0  751814.0  414398.0  1490228.0  779648.0  624738.0   

RegionName     78719     78742  
01/31/2000  144311.0  117246.0  
02/29/2000  145057.0  117515.0  
03/31/2000  144970.0  116911.0  
04/30/2000  145948.0  116133.0  
05/31/2000  146070.0  114755.0  
...              ...       ...  
02/28/2022  415204.0  424746.0  
03/31/2022  427896.0  434459.0  
04/30/2022  439443.0  441220.0  
05/31/2022  448147.0  447581.0  
06/30/2022  443553.0  443758.0  

[270 rows x 43 columns]

In [61]:
# Plot Zillow data for all zip codes

dates_df.hvplot.line(rot=90, 
    xlabel="Dates",
    ylabel="Average Home Prices",
    width=2500, 
    height=600
).opts(yformatter="%.0f")

:NdOverlay   [RegionName]
   :Curve   [index]   (value)

In [62]:
# Parse through and extract the Dec 2007 - June 2010 data 

recession_dates = dates_df.loc["11/30/2007":"06/30/2010", :]
display(recession_dates.head())

RegionName     78704     78745     78748     78758     78741     78753  \
11/30/2007  373165.0  189872.0  203840.0  156372.0  181694.0  159376.0   
12/31/2007  374997.0  190459.0  204525.0  156577.0  182941.0  159997.0   
01/31/2008  375049.0  190528.0  204625.0  156308.0  183026.0  160229.0   
02/29/2008  374256.0  190409.0  204437.0  155932.0  182415.0  160210.0   
03/31/2008  374282.0  190847.0  204827.0  155928.0  182422.0  160065.0   

RegionName     78759     78744     78723     78749  ...     78756     78721  \
11/30/2007  313919.0  152706.0  213829.0  240005.0  ...  362903.0  184876.0   
12/31/2007  315961.0  153375.0  214482.0  241460.0  ...  363836.0  185562.0   
01/31/2008  317099.0  153869.0  213949.0  242065.0  ...  365390.0  184115.0   
02/29/2008  318336.0  153644.0  214008.0  242037.0  ...  365745.0  183040.0   
03/31/2008  319758.0  153266.0  214708.0  242067.0  ...  366940.0  182252.0   

RegionName     78730     78736     78725     78733     78722     78652  \
11/30/2007  584038.0  273631.0  146513.0  534944.0  286480.0  232144.0   
12/31/2007  584989.0  274480.0  147557.0  535289.0  287711.0  234010.0   
01/31/2008  584975.0  274285.0  147759.0  533482.0  288514.0  235082.0   
02/29/2008  585169.0  274090.0  147375.0  531920.0  288146.0  234806.0   
03/31/2008  587389.0  273752.0  146900.0  531571.0  287985.0  236381.0   

RegionName     78719     78742  
11/30/2007  157543.0  152639.0  
12/31/2007  159036.0  153429.0  
01/31/2008  160748.0  153163.0  
02/29/2008  161033.0  153271.0  
03/31/2008  159892.0  153022.0  

[5 rows x 43 columns]

In [63]:
recession_dates.hvplot.line(rot=90, 
    xlabel="Dates",
    ylabel="Average Home Prices",
    width=2500, 
    height=600
).opts(yformatter="%.0f")

:NdOverlay   [RegionName]
   :Curve   [index]   (value)

In [64]:
days_elapsed = [num for num in np.arange(0, 30, 1)]

    # in and out sample testing
    # inflation, certain president, migration, global pandemic, interest rates 
    # spin the story; If i'm going to buy now, why would this model work now

    # Extract prices that are not NaN from previous calculations
experimental_prices = [x for x in exp_prices if np.isnan(x) == False]

# Pull the last row of data corresponding to 6/30/2022 from original dataframe for comparison and clean the indices
data_prices = dates_df[zipcode_updated].iloc[-1]
data_prices.index = [num for num in np.arange(0, 37, 1)]

# Clean indices of zip code list
zipcode_updated.index = [num for num in np.arange(0, 37, 1)]

In [65]:
# Merge zip codes and models into one dataframe

zip_fits_df = pd.DataFrame({
    'Zip Code': zipcode_df,
    'Slope': slope,
    'Intercept': intercept
})

zip_fits_df.dropna(inplace=True)

display(len(zip_fits_df))
display(zip_fits_df.head())

# Create an updated zip code dataframe with only zip codes that don't have NaN

zipcode_updated = zip_fits_df['Zip Code']



ValueError: array length 49 does not match index length 43

## Linear Regression Fitting to Data

In [71]:
# Apply linear regression from numpy to the dataset

months_elapsed = [num for num in np.arange(0, 960, 30)]

zip_fits = []
for (column_name, column_data) in recession_dates.iteritems():
    try:
        current_fit = np.polyfit(months_elapsed, recession_dates[column_name], 1)
        zip_fits.append(current_fit)
    except: 
        zip_fits.append(np.nan)
    finally: 
        continue

display(zip_fits)

[array([-4.09964260e+01,  3.81202932e+05]),
 array([-4.75259042e+00,  1.91203205e+05]),
 array([-7.95028104e+00,  2.04882818e+05]),
 array([-8.95136852e+00,  1.56383574e+05]),
 array([-2.76902859e+01,  1.85082358e+05]),
 array([-8.05429497e+00,  1.59240528e+05]),
 array([-2.19062256e+01,  3.22236364e+05]),
 array([-1.20281281e+01,  1.53444330e+05]),
 array([-1.0920534e+01,  2.1739033e+05]),
 array([-1.27989431e+01,  2.41718352e+05]),
 array([-1.49500244e+00,  2.01084426e+05]),
 array([-1.43293072e+01,  2.67401784e+05]),
 array([-3.69755621e+00,  1.98208551e+05]),
 array([-4.47920516e+01,  5.02081585e+05]),
 array([-3.41478067e+01,  2.41859449e+05]),
 array([-8.75610337e+00,  1.89319932e+05]),
 array([-5.06887891e+01,  7.74490506e+05]),
 array([-8.09461755e+01,  6.93138972e+05]),
 array([-6.34714687e+00,  2.80708767e+05]),
 array([-2.08873228e+01,  2.41930574e+05]),
 array([-1.64569098e+01,  2.85606494e+05]),
 array([-4.06347446e+01,  4.10985500e+05]),
 array([-2.38272727e+01,  1.962626

In [67]:
# Parse through the arrays and build 2 lists, one for the intercept and the other for the slope

linear_intercept = []
linear_slope = []

for arr in zip_fits:
    try:
        linear_slope.append(arr.tolist()[0])
        linear_intercept.append(arr.tolist()[1])
    except: 
        linear_slope.append(np.nan)
        linear_intercept.append(np.nan)
    finally: 
        continue

display(linear_intercept)
display(linear_slope)

[381202.93181818206,
 191203.2045454546,
 204882.8181818183,
 156383.57386363638,
 185082.35795454553,
 159240.52840909094,
 322236.3636363637,
 153444.32954545453,
 217390.32954545465,
 241718.35227272732,
 201084.4261363637,
 267401.7840909091,
 198208.55113636365,
 502081.5852272729,
 241859.44886363635,
 189319.93181818185,
 774490.5056818184,
 693138.9715909092,
 280708.76704545453,
 241930.57386363635,
 285606.4943181818,
 410985.50000000006,
 196262.61931818188,
 327895.45454545453,
 178343.32954545456,
 147987.5625,
 203483.65340909097,
 154060.48863636368,
 360848.72727272735,
 359962.81818181823,
 364336.6477272727,
 368899.1420454546,
 445759.6420454547,
 369375.04545454547,
 182853.26136363635,
 594651.7727272728,
 275639.4090909091,
 146884.19886363638,
 536666.0000000001,
 289561.7045454546,
 236311.77272727282,
 159172.23863636368,
 153147.25000000006]

[-40.99642595307934,
 -4.75259042033232,
 -7.950281036168187,
 -8.951368523949217,
 -27.69028592375374,
 -8.054294965786907,
 -21.906225562072454,
 -12.028128054740955,
 -10.920533968719516,
 -12.79894305962862,
 -1.4950024437928116,
 -14.329307184750729,
 -3.6975562072336103,
 -44.79205156402748,
 -34.14780669599212,
 -8.756103372434032,
 -50.688789100684254,
 -80.94617546432067,
 -6.3471468719452275,
 -20.88732282502443,
 -16.45690982404691,
 -40.63474462365588,
 -23.827272727272728,
 -32.41663611925707,
 -5.8186522482893475,
 -15.701075268817183,
 -23.113569159335373,
 -21.31872556207231,
 -32.31056940371455,
 -27.448130498533747,
 -72.28486070381233,
 -38.129472140762424,
 -57.903195259042164,
 -23.68907624633422,
 -35.84989002932547,
 -66.64850317693065,
 -25.442210410557184,
 -11.14437927663736,
 -55.065860215053775,
 -22.63236192570871,
 -13.934054252199534,
 -24.983913734115447,
 -24.431250000000034]

In [68]:
# Create a dataframe including the slope and intercept of the linear regression models for qualifying zip codes

zip_linear_fits_df = pd.DataFrame({
    'Zip Code': zipcode_df,
    'Slope': linear_slope,
    'Intercept': linear_intercept
})

zip_linear_fits_df.dropna(inplace=True)

zip_linear_fits_df.index = [num for num in np.arange(0, len(zip_linear_fits_df), 1)]

display(len(zip_linear_fits_df))
display(zip_linear_fits_df.head())
display(zip_linear_fits_df.tail())

43

Zip Code      Slope      Intercept
0     78704 -40.996426  381202.931818
1     78745  -4.752590  191203.204545
2     78748  -7.950281  204882.818182
3     78758  -8.951369  156383.573864
4     78741 -27.690286  185082.357955

Zip Code      Slope      Intercept
38     78733 -55.065860  536666.000000
39     78722 -22.632362  289561.704545
40     78652 -13.934054  236311.772727
41     78719 -24.983914  159172.238636
42     78742 -24.431250  153147.250000

## Validating/Testing Linear Regression Model

In [77]:
# Create new dataframe that includes the experimental and actual data values along with zip codes

last_three_months = [num for num in np.arange(0, 120, 30)]

actual_values_df = dates_df[zip_linear_fits_df['Zip Code']].iloc[-3:]

experimental_values_df = pd.DataFrame()
experimental_values_df.index = ['04/30/2022', '05/31/2022', '06/30/2022']

num = 0
for model in linear_slope:
    if model == np.nan:
        continue
    else:
        april_price = round(linear_intercept[num] + model * last_three_months[0], 2)
        may_price = round(linear_intercept[num] + model * last_three_months[1], 2)
        june_price = round(linear_intercept[num] + model * last_three_months[2], 2)
        experimental_values_df[zip_linear_fits_df['Zip Code'].iloc[num]] = [april_price, may_price, june_price]
        num+=1


# Calculate percent change between experimental and actual data values and convert to series 

display(experimental_values_df)
display(actual_values_df)

# linear_percent_change = (data_prices - linear_prices)/data_prices * 100
# linear_percent_change = pd.Series(linear_percent_change)
# display(linear_percent_change.head())
# display(linear_percent_change.tail())


78704      78745      78748      78758      78741      78753  \
04/30/2022  381202.93  191203.20  204882.82  156383.57  185082.36  159240.53   
05/31/2022  379973.04  191060.63  204644.31  156115.03  184251.65  158998.90   
06/30/2022  378743.15  190918.05  204405.80  155846.49  183420.94  158757.27   

                78759      78744      78723      78749  ...      78756  \
04/30/2022  322236.36  153444.33  217390.33  241718.35  ...  369375.05   
05/31/2022  321579.18  153083.49  217062.71  241334.38  ...  368664.37   
06/30/2022  320921.99  152722.64  216735.10  240950.42  ...  367953.70   

                78721      78730      78736      78725      78733      78722  \
04/30/2022  182853.26  594651.77  275639.41  146884.20  536666.00  289561.70   
05/31/2022  181777.76  592652.32  274876.14  146549.87  535014.02  288882.73   
06/30/2022  180702.27  590652.86  274112.88  146215.54  533362.05  288203.76   

                78652      78719      78742  
04/30/2022  236311.77  159172.24  153147.25  
05/31/2022  235893.75  158422.72  152414.31  
06/30/2022  235475.73  157673.20  151681.38  

[3 rows x 43 columns]

RegionName      78704     78745     78748     78758     78741     78753  \
04/30/2022  1087691.0  612028.0  561694.0  561618.0  515889.0  494108.0   
05/31/2022  1096965.0  618316.0  568101.0  567226.0  519962.0  500128.0   
06/30/2022  1095829.0  613862.0  563042.0  560180.0  514218.0  495392.0   

RegionName     78759     78744     78723     78749  ...     78756     78721  \
04/30/2022  876764.0  460693.0  670097.0  713245.0  ...  952069.0  576838.0   
05/31/2022  884352.0  467861.0  672464.0  721249.0  ...  957718.0  579854.0   
06/30/2022  874722.0  464934.0  661809.0  713630.0  ...  950845.0  570825.0   

RegionName      78730     78736     78725      78733     78722     78652  \
04/30/2022  1421492.0  748991.0  407091.0  1465264.0  773833.0  624577.0   
05/31/2022  1447688.0  757909.0  415376.0  1487102.0  787315.0  632163.0   
06/30/2022  1443527.0  751814.0  414398.0  1490228.0  779648.0  624738.0   

RegionName     78719     78742  
04/30/2022  439443.0  441220.0  
05/31/2022  448147.0  447581.0  
06/30/2022  443553.0  443758.0  

[3 rows x 43 columns]

In [84]:
percent_change = (experimental_values_df - actual_values_df)/actual_values_df * 100

display(percent_change)

RegionName       78704       78745       78748       78758       78741  \
04/30/2022  185.331228  220.092969  174.153782  259.128520  178.734829   
05/31/2022  188.695482  223.622925  177.604102  263.338495  182.202086   
06/30/2022  189.333022  221.531673  175.453045  259.443450  180.348580   

RegionName       78753       78759       78744       78723       78749  ...  \
04/30/2022  210.290351  172.087234  200.234619  208.246002  195.072757  ...   
05/31/2022  214.548088  175.002878  205.624728  209.801716  198.858787  ...   
06/30/2022  212.043663  172.565305  204.430306  205.353863  196.172964  ...   

RegionName       78756       78721       78730      78736       78725  \
04/30/2022  157.751302  215.464980  139.046123  171.72856  177.150980   
05/31/2022  159.780461  218.990618  144.272730  175.72746  183.436621   
06/30/2022  158.414306  215.892545  144.395160  174.27168  183.415839   

RegionName       78733       78722       78652       78719       78742  
04/30/2022  173.030898  167.242871  164.302112  176.080176  188.101811  
05/31/2022  177.955707  172.537926  167.986329  182.880511  193.660746  
06/30/2022  179.402706  170.519718  165.308871  181.311599  192.559311  

[3 rows x 43 columns]

In [ ]:
# Create a new dataframe with experimental values and actual data values along with the percent difference between the two

linear_comparison_df = pd.DataFrame({
    'Zip Code': zipcode_updated,
    'Experimental Value ($)': exp_linear_prices,
    'Accepted/Data Value ($)': data_prices
})

linear_comparison_df['Delta (%)'] = linear_percent_change

display(linear_comparison_df.head(15))
display(linear_comparison_df.tail())

Zip Code  Experimental Value ($)  Accepted/Data Value ($)  Delta (%)
0      78704           724834.473276                1095829.0  33.855148
1      78745           389827.319510                 613862.0  36.495936
2      78748           359643.390224                 563042.0  36.124944
3      78758           341246.534572                 560180.0  39.082699
4      78741           330927.406747                 514218.0  35.644531
5      78753           306815.886218                 495392.0  38.066039
6      78759           563460.853669                 874722.0  35.584008
7      78744           280628.809776                 464934.0  39.641151
8      78723           454283.939033                 661809.0  31.357244
9      78749           453447.864243                 713630.0  36.458968
10     78729           381061.892138                 605979.0  37.116321
11     78727           397681.060085                 625401.0  36.411829
12     78731           856235.998485                1342438.0  36.217837
13     78705           344581.257717                 442315.0  22.095959
14     78728           344181.515324                 544132.0  36.746687

Zip Code  Experimental Value ($)  Accepted/Data Value ($)  Delta (%)
32     78725           242080.109239                 414398.0  41.582703
33     78733           900176.093983                1490228.0  39.594740
34     78652           403573.170233                 624738.0  35.401213
35     78719           263305.151838                 443553.0  40.637274
36     78742           278757.970976                 443758.0  37.182435

# **Conclusion**

### It can be concluded that for this data set within this time period of January 31, 2000 to June 30, 2022, neither exponential nor linear regression can accurately model and predict current and near-future values. With differences of -1.8 to 2.7 between exponential and linear percent changes, it cannot be concluded which models the data more accurately. Additionally, with percent changes mostly falling between 30-40% for both exponential and linear models, neither can be used for home price prediction and another model/regression should be used for prediction purposes.

### Therefore we can conclude that the data is, obviously, not linear, but is also not exponential as well. Even popular zip codes such as 78737 and 78739 which have the highest spike in home prices cannot be modeled as exponential. This can be explained by the fact that whereas most conventional exponential models exhibit radical increases towards the tail end of graphs, these two neighborhoods exhibited gradual increases in home prices since 2012 before it saw a dramatic spike in home prices in the beginning of 2021.

### Upon further analysis and retrospection, neither linear nor exponential models would have worked because the linear regression would have been heavily effected by dramatic increase in prices seen in 2021-2022, while the exponential regression would have been effected by a general increase in home prices seen between 2012-2021. Most neighborhoods that did not exhibit too much of an increase between 2012-2021 also did not see the dramatic increases in 2021-2022 compared to zip codes such as 78737 and 78739, but regardless it's effects were felt in the regression models.

from https://www.theatlantic.com/business/archive/2010/07/how-texas-is-dominating-the-recession/60721/

2. Stable Real Estate
Real estate executives and economists struggled to find one reason why the Texas economy largely avoided the real estate boom and bust, but a few theories emerged. First, San Antonio Mayor Julian Castro suggested that a reliance on property taxes in Texas (compared to California) might have dulled real estate appreciation. Second, the banks that survived the Savings and Loan crisis in the 1980s have mostly held onto conservative and un-exotic lending practices. Third, land and utilities are generally cheaper throughout Texas, which holds down the cost of the living. Fourth, besides Dallas, Texas' major cities have diversified away from the kind of real estate and financial services addiction that plagued CaliFlAriVada (that's CA, FL, AZ, NV), where the recession has been the most severe.